DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,290 @@
+            IRegistrationDataService registrationDataService,
+            IUserService userService
+        )
+        {
+            this.userDataService = userDataService;
+            this.registrationDataService = registrationDataService;
+            this.jobGroupsDataService = jobGroupsDataService;
+            this.userService = userService;
+        }
+
+        public BulkUploadResult ProcessDelegatesFile(IFormFile file, int centreId, DateTime? welcomeEmailDate)
+        {
+            var table = OpenDelegatesTable(file);
+            return ProcessDelegatesTable(table, centreId, welcomeEmailDate);
+        }
+
+        public IXLTable OpenDelegatesTable(IFormFile file)
+        {
+            var workbook = new XLWorkbook(file.OpenReadStream());
+            var worksheet = workbook.Worksheet(DelegateDownloadFileService.DelegatesSheetName);
+            var table = worksheet.Tables.Table(0);
+
+            if (!ValidateHeaders(table))
+            {
+                throw new InvalidHeadersException();
+            }
+
+            return table;
+        }
+
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate)
+        {
+            var jobGroupIds = jobGroupsDataService.GetJobGroupsAlphabetical().Select(item => item.id).ToList();
+            var delegateRows = table.Rows().Skip(1).Select(row => new DelegateTableRow(table, row)).ToList();
+
+            foreach (var delegateRow in delegateRows)
+            {
+                ProcessDelegateRow(centreId, welcomeEmailDate, delegateRow, jobGroupIds);
+            }
+
+            return new BulkUploadResult(delegateRows);
+        }
+
+        private void ProcessDelegateRow(
+            int centreId,
+            DateTime? welcomeEmailDate,
+            DelegateTableRow delegateRow,
+            IEnumerable<int> jobGroupIds
+        )
+        {
+            if (!delegateRow.Validate(jobGroupIds))
+            {
+                return;
+            }
+
+            var delegateUserByCandidateNumber =
+                GetDelegateUserByCandidateNumberOrDefault(centreId, delegateRow.CandidateNumber);
+
+            if (!string.IsNullOrEmpty(delegateRow.CandidateNumber) && delegateUserByCandidateNumber == null)
+            {
+                delegateRow.Error = BulkUploadResult.ErrorReason.NoRecordForDelegateId;
+                return;
+            }
+
+            var delegateUserByAliasId = GetDelegateUserByAliasIdOrDefault(centreId, delegateRow.AliasId);
+
+            if (delegateUserByAliasId != null && delegateUserByCandidateNumber != null &&
+                delegateUserByAliasId.CandidateNumber != delegateUserByCandidateNumber.CandidateNumber)
+            {
+                delegateRow.Error = BulkUploadResult.ErrorReason.AliasIdInUse;
+                return;
+            }
+
+            // At this point if both delegateUserByCandidateNumber and delegateUserByAliasId are not null

```

Comment:
Having to have comments to describe what your code does is usually a sign of there being potential for refactor. I feel like in this case it would be easier to follow if we did something along the lines of:

```
var userToUpdate = delegateUserByCN ?? delegateUserByAlias;
if (userToUpdate == null) {
    // Register
} else {
   ProcessPotentialUpdate(centreId, delegateRow, userToUpdate);
}
```